tasks/init.rb

```diff
@@ -2,41 +2,28 @@
 
 require 'puppet'
 require 'json'
-require 'open3'
-
-def get(fact)
-  if Gem.win_platform?
-    require 'win32/registry'
-    begin
-      installed_dir = Win32::Registry::HKEY_LOCAL_MACHINE.open('SOFTWARE\Puppet Labs\Puppet') do |reg|
-        dir = reg['RememberedInstallDir64']
-        break dir if File.exist?(dir)
-        reg['RememberedInstallDir']
-      end
-      facter = File.join(installed_dir, 'bin', 'facter.bat')
-    rescue Win32::Registry::Error
-      facter = ''
-    end
-  else
-    facter = '/opt/puppetlabs/puppet/bin/facter'
-  end
-
-  # Fall back to PATH lookup if puppet-agent isn't installed
-  facter = 'facter' unless File.exist?(facter)
-
-  cmd = [facter, '-p', '--json']
-  cmd << fact if fact
-  stdout, stderr, status = Open3.capture3(*cmd)
-  raise stderr if status != 0
-  stdout
-end
 
 params = JSON.parse(STDIN.read)
 fact = params['fact']
 
 begin
-  result = get(fact)
-  puts result
+  Puppet.initialize_settings
+  $LOAD_PATH << Puppet[:libdir] unless $LOAD_PATH.include?(Puppet[:libdir])
+  Facter.reset
+  Facter.search_external([Puppet[:pluginfactdest]])
+  if Puppet.respond_to? :initialize_facts
+    Puppet.initialize_facts
+  else
+    Facter.add(:puppetversion) do
+      setcode { Puppet.version.to_s }
+    end
+  end
+
+  if fact
+    puts({ fact => Facter.value(fact) }.to_json)
+  else
+    puts(Facter.to_hash.to_json)
```

Comment:
The facter terminus in puppet does some work to setup external search paths. Is that needed here in order to provide the `-p` functionality?